### PR TITLE
Remove deprecated commands Implicit Arguments and Arguments Scope

### PR DESCRIPTION
**Pull request by jashug**
**Kind:** cleanup

The CHANGES file shows these were added for 8.4, and git blame shows they have been deprecated for at least 6 years. (Though apparently the deprecation warning was only visible from 8.6) The code removal is small, but I think not insignificant.

https://github.com/coq/coq/blob/8dd6d091ffbfa237f7266eeca60187263a9b521f/CHANGES#L1411-L1413

Some notes:
1. `Implicit Arguments foo [].` becomes `Arguments foo : clear implicits.`
`Implicit Arguments foo.` becomes `Arguments foo : default implicits.`
Because `Arguments` works by place and not by name, some things had to be rearranged and padded with spacers.

2. `Arguments Scope` was able to specify more scopes than `Arguments` can see rigid parameters. This could be used to give scopes after coercion to `Funclass` or after a computed type. Use the flag `extra scopes` to do the same thing with `Arguments`.

The first commit has all the changes to the test suite, and the second has actual code removal.

Edit: after a rebase, not sure where to put the changes entry right now, so I'll leave my proposed text here:
```
- Removed deprecated commands Arguments Scope and Implicit Arguments
  (not the option). Use the Arguments command instead.
```

### CODE REVIEW

**Comment by jashug** [comment]
It looks like until 2 years ago the deprecation warning was given with
`Flags.if_verbose Feedback.msg_warning`. Would that not be printed normally in CoqIDE and compiling?
I think the change was in by 8.6, but I'm not sure.

My travis gives lots of tracked developments that need to be updated:
https://travis-ci.org/jashug/coq/builds/344206028

I'll work on sending PRs there to help modernize.

- [ ] Fix flocq (commit exists in middle of feature branch) -> compcert
- [ ] Fix math-comp = ssreflect ([issue exists](https://github.com/math-comp/math-comp/issues/93)) -> iris-rust, coquelicot?, vst?
- [ ] Fix fiat-crypto
- [ ] Fix math-classes -> formal-topology
- [ ] Fix sf

**Comment by maximedenes** [comment]
I already started doing it for mathcomp, so if you want to take over, I can try to submit a partial patch, so that you don't have to redo the work.

**Comment by jashug** [comment]
That would be great! If you could push a branch to your fork, I can work off that.

**Comment by maximedenes** [comment]
My patch is much smaller than I remembered, I wonder if I haven't stored a more complete version somewhere else, but couldn't find it. So here it is:

https://github.com/maximedenes/math-comp/tree/fix-warnings

Note that it does not just fix deprecated `Arguments` variants, it also avoid relying on some compatibility notation from stdlib.

**Comment by jashug** [comment]
Link to flocq commit in feature branch: https://gitlab.inria.fr/flocq/flocq/commit/78e75ff116752a497343fc0cbb91c060a7d79a0a#note_76147.
When I asked about cherry-picking that commit, I was told

> The master branch supports Coq versions as old as 8.4. Unfortunately, these versions provide a fairly broken Arguments command, so this commit cannot be cherry-picked onto master.

**Comment by Zimmi48** [comment]
Answering here to the thread between @jashug @maximedenes and @silene because I don't want to create an account over there: if the master branch needs to stay compatible with 8.4, it shouldn't prevent us from moving on. We can just decide to track the feature branch instead in our CI.

**Comment by herbelin** [comment]
>  Arguments Scope was able to specify more scopes than Arguments can see rigid parameters. This could be used to give scopes after coercion to Funclass or after a computed type. I couldn't find a way to replicate this functionality with Arguments, so I removed tests in success/Scopes.v and bugs/closed/4865.v, and deleted bugs/closed/2667.v entirely.

Maybe it is not so difficult to fix `Arguments` to handle this. After all, iirc, `Arguments` is eventually calling the same code as `Arguments Scope` behind the scene, so the infrastructure to deal with the extra arguments is still there.

File: test-suite/bugs/closed/2667.v
Comment (JasonGross) [inline test-suite/bugs/closed/2667.v]: Can't you do this with
```coq
Arguments Scall _ _%Cminor.
```
(You might need a ` Delimit Scope` first)

File: test-suite/bugs/closed/4865.v
Comment (JasonGross) [inline test-suite/bugs/closed/4865.v]: Does
```coq
Arguments lam _%nat _%nat.
```
not work?

**Review by JasonGross — DISMISSED** [review]
If ` Arguments` cannot do what ` Argument Scopes` could, why are we removing ` Argument Scopes`?

File: test-suite/bugs/closed/2667.v
Comment (jashug) [reply test-suite/bugs/closed/2667.v]: That fails with `Extra arguments: _.`, but `Arguments Scall _ _%Cminor : extra scopes` works!

This is an undocumented flag to `Arguments` that does what we needed. Rather than removing these tests, I'll update to document the flag.

**Comment by jashug** [comment]
So I just found an undocumented flag, `Arguments foo : extra scopes` that does what I was missing. I'll update to document this feature.

**Comment by jashug** [comment]
My updated understanding of the timeline of various things:
* 8.4: Arguments command added, Implicit Arguments and Arguments Scope deprecated, but no warning printed except in verbose mode.
* 8.6: Arguments semantics changed closer to the current semantics, warning printed.
So the deprecated commands should probably be considered as compatibility for 8.5, not 8.3 as I originally though.

Since flocq is currently supporting 8.4, and math-comp is currently supporting 8.5, it may be too soon to remove these commands. At least having this in 8.8 would feel very rushed, and after 8.8 is relesed some of these external projects may decide to drop support for 8.5.

**Comment by herbelin** [comment]
> So I just found an undocumented flag, Arguments foo : extra scopes that does what I was missing. I'll update to document this feature.

I learn everyday. Wonderful!

**Comment by jashug** [comment]
What's the procedure for requesting changes / submitting pull requests to the Software Foundations books?

**Comment by maximedenes** [comment]
You can directly send a patch by e-mail to Benjamin Pierce. The book is going to be migrated to a collaborative platform.

**Comment by jashug** [comment]
Summary of current state of affairs: patch was sent to Software Foundations and acknowledged, but not yet updated.
math-comp and flocq appear to be waiting for 8.8 to be released.
compcert, iris-rust, coquelicot, vst are dependent on either math-comp or flocq, and hopefully will work once fixes are merged upstream.

**Comment by Zimmi48** [comment]
CompCert depends on Flocq but IIUC its sources contain a copy of Flocq. So you might still try to submit your patch to them, especially given that they already dropped support for versions < 8.6.1.

**Review by gares — APPROVED** [review]

**Comment by herbelin** [comment]
Meta-questions:
- when written "@jashug requested review from ejgallego, gares, herbelin, letouzey, mattam82, maximedenes and ppedrot as code owners 22 hours ago", it shall be understood as "the robot requested reviews from ..." based on the contents of the `CODEOWNERS` file, right? (Or, @jashug, did you really yourself explicitly asked for reviews from 7 people?)
- How do we proceed? For instance, @gares apparently looked at it and checked what I imagine he estimated he had to check (@gares, you did make a pass over the whole PR and changes in the test-suite already I guess?). To share work, I'd then be willing on my side to trust @gares' review and to simply formally validate the review required from me. Is that ok for anyone? (Or does someone prefer that I make a pass myself on all the details of the PR?)
- Also, with respect to the different parts of a review, like regression tests, documentation, etc. as mentioned by @Zimmi48 recently, how do we proceed? Do we follow a standard check-list about the different points to check?

**Comment by jashug** [comment]
@herbelin 
> when written "@jashug requested review from ejgallego, gares, herbelin, letouzey, mattam82, maximedenes and ppedrot as code owners 22 hours ago", it shall be understood as "the robot requested reviews from ..." based on the contents of the CODEOWNERS file, right? (Or, @jashug, did you really yourself explicitly asked for reviews from 7 people?)

Yes, that was the github robot. I think the clue phrase there is "as code owners", but also I don't have the permissions to request a review from anyone.

Status update:
* math-comp was updated to use Arguments.
* Software Foundations hasn't yet released a new version containing the patch I sent.
* flocq: still waiting for flocq 3.0.

**Comment by Zimmi48** [comment]
This is a case of a large PR where many components are touched but not all reviews are really required. I think the review from @gares is likely more than enough. However, it'd be better if @gares could confirm this.

**Comment by jashug** [comment]
I created a patch for flocq that should be compatible with 8.4: see https://gitlab.inria.fr/flocq/flocq/issues/1.
@maximedenes Are there fundamental reasons this is a bad idea?

**Comment by gares** [comment]
@Zimmi48 I'll take care of this PR.

@jashug thanks for your excellent job!
We don't need the upstream authors of all the projects we run via CI to merge your PRs before merging this one. We just need your PRs to exist and be registered as user overlays: see for example this file: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/06676-gares-proofview-goals-come-with-a-state.sh . In this way our CI can test the full package of PRs (and I can just look at the CI summary to know if this PR is ready).
You can add one overlay file named after this PR to collect all your external PRs.
I will then take care of pinging the upstream authors when this PR is merged, so that your others PRs get eventually merged upstream.

**Comment by Zimmi48** [comment]
> We don't need the upstream authors of all the projects we run via CI to merge your PRs before merging this one.

We don't need the upstream authors of all the projects we run via CI to merge your PRs before *testing* this one.

**Comment by Zimmi48** [comment]
Sorry for this. Bad manipulation on my phone.

**Comment by Zimmi48** [comment]
> I will then take care of pinging the upstream authors when this PR is merged, so that your others PRs get eventually merged upstream.

When the patches are backward compatible (i.e. most patches to libraries and not plugins), we ensure that the patches are merged *before* the main PR is merged to ensure we don't break our CI for no good reason. Indeed, overlays are made to work only during the PR lifetime.

**Comment by Zimmi48** [comment]
I'll try to improve the merging doc to explain this better.

**Comment by gares** [comment]
> I'll try to improve the merging doc to explain this better.

Please do, because this *before* thing is new to me.

**Comment by jashug** [comment]
@gares When the external project is on github, it seems relatively straight-forward to fork and set that as the overlay. For projects like flocq or software foundations which aren't on github, this (seems) not so straightforward. Since all the changes needed should be backwards compatible, and since this PR probably shouldn't be merged until all external projects have been updated, I decided making overlays would be more work than benefit.

**Comment by Zimmi48** [comment]
For SF there is indeed no good way to make overlays. For the rest of the external developments, since they are on git, you can always clone and push to a new repository on GitHub. Then, you can always prepare a "manual" PR by opening a new issue and saying: please pull from here.

**Comment by jashug** [comment]
Closing and reopening to run CI.
@gares I think this is ready to go.

**Comment by jashug** [comment]
Or rather I'll add the changes entry...

**Review by ejgallego — APPROVED** [review]
I have nothing else to say than thanks to @jashug .

**Pull request merged**
